### DISCUSSION TITLE
N+1 queries with meilisearch

### DISCUSSION THREAD

**Discussion by coder2000** [ddnexus/pagy#400] [Q&A] [answered]
Don't know if this is pagy or meilisearch specific but looking to limit N+1 queries when executing searches. Where would I add the required includes?

**Comment by ddnexus** [comment]
Why your intention to limit N+1 queries would have anything to do with where to add the required includes?

**Reply by coder2000** [reply]
When loading related models to reduce N+1 you usually use `Model.includes(:relation)` do you not? How would I do that with pagy and meilisearch?

**Reply by coder2000** [reply]
We sort of discussed this in the past when I opened an issue about it and there was a workaround we came up with but obviously it isn't very intuitive.

**Reply by ddnexus** [reply]
Oh... that "includes". Sorry, I was still sleepy. I was thinking about the includes of pagy modules into your app. 🤦‍♂️

I've found your Gitter messages. I have no direct experience with meilisearch, so I don't know whether there is a better way to do it.

Since avoiding extra queries means having the data included in the collection, the most straightforward and performant way that comes to my mind is including the data in the index. It might not be optimal for other things that I may ignore though.

**Reply by ddnexus** [reply]
@brunoocasali Maybe you have some official way to tackle this?

**Comment by brunoocasali** [comment, marked as answer]
Hi @coder2000 I think we discussed that in the Meilisearch repo https://github.com/meilisearch/meilisearch-rails/issues/133#issuecomment-1062966592

I'm not sure where I misread the discussion, but I thought you understood where to put the `includes`. Also, you must add the `ActiveRecord_Relation.include Pagy::Meilisearch` in the model.

```rb
class Restaurant < ApplicationRecord
  has_many :reviews, dependent: :destroy

  include MeiliSearch::Rails
  extend Pagy::Meilisearch
  ActiveRecord_Relation.include Pagy::Meilisearch # <--- this line

  meilisearch ...
end
```
Then you can 

```rb
restaurants = Restaurant.includes(:reviews).pagy_search(params[:query])
@pagy, @restaurants = pagy_meilisearch(restaurants, items: 25)
```

Are you still having trouble with the N+1?

**Reply by coder2000** [reply]
No. This was a reply from someone else.

**Reply by brunoocasali** [reply]
Nice, I think you can accept my answer @ddnexus and close this discussion :)
